### PR TITLE
Feature admin interface with pagination for users and programmes

### PR DESCRIPTION
**Pull request by marceldobrica**
Complete admin interface but without:
1. Pagination - now I have pagination on users and programmes - enough for moment.
2. More validations - still to do.
3. Valid trainers in select box - still to do.
4. Set customers in programmes.... or a special admin page to set customers for programmes.

### CODE REVIEW

File: src/Controller/Admin/BuildingsController.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        $buildings = $this->buildingRepository->findAll();
+
+        return $this->render('admin/buildings/index.html.twig', [
+            'buildings' => $buildings
+        ]);
+    }
+
+    /**
+     * @Route ("admin/buildings/add", name="app_admin_buildings_add", methods={"GET", "POST"})
```
Comment (Raresmldvn) [inline src/Controller/Admin/BuildingsController.php]: you don't need the `/add` here and you can also call it just `/building` without the s

File: src/Controller/Admin/BuildingsController.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        $building = $this->buildingRepository->findOneBy(['id' => $id]);
+        $form = $this->createForm(DeleteCancelType::class);
+        $form->handleRequest($request);
+        if ($form->isSubmitted() && $form->isValid()) {
+            if ($form->get('delete')->isClicked()) {
+                $this->entityManager->remove($building);
+                $this->entityManager->flush();
+                $this->addFlash(
+                    'warning',
```
Comment (Raresmldvn) [inline src/Controller/Admin/BuildingsController.php]: I don't think a sucessful delete is a warning

File: src/Controller/Admin/UsersController.php [hunk 3]
```diff
@@ -2,29 +2,148 @@
...
+            /** @var User $user */
+            $user = $form->getData();
+            $user->setPassword($this->passwordHasher->hashPassword($user, $user->getPassword()));
+            $this->entityManager->persist($user);
+            $this->entityManager->flush();
+            $this->addFlash(
+                'success',
+                'You have created a new user!'
+            );
+
```
Comment (Raresmldvn) [inline src/Controller/Admin/UsersController.php:83]: remove this unnecessary newline

File: src/Entity/Building.php [hunk 4]
```diff
@@ -64,4 +73,9 @@ public static function createFromDto(BuildingDto $buildingDto): self
 
         return $building;
     }
+
+    public function __toString()
```
Comment (Raresmldvn) [inline src/Entity/Building.php]: return type string

Comment (Raresmldvn) [reply src/Entity/Building.php]: and for all the other `_toString` methods as well

File: src/Form/Type/BuildingType.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class BuildingType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $builder
+            ->add('address')
+            ->add('startTime')
+            ->add('endTime')
```
Comment (Raresmldvn) [inline src/Form/Type/BuildingType.php]: you could add some types here as well, to enforce validations and rendering

File: src/Form/Type/ProgrammeType.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        $builder
+            ->add('name')
+            ->add('description')
+            ->add('startDate')
+            ->add('endDate')
+            ->add('isOnline')
+            ->add('maxParticipants')
+            ->add('trainer')
+            ->add('room')
+            ->add('customers')
```
Comment (Raresmldvn) [inline src/Form/Type/ProgrammeType.php]: types here?

File: src/Repository/UserRepository.php [hunk 7]
```diff
@@ -39,4 +39,29 @@ public function remove(User $entity, bool $flush = true): void
...
     }
+
+    public function findAllPaginated($currentPage, $pageSize): array
+    {
+        $currentPosition = ($currentPage - 1) * $pageSize;
+        $query = $this->_em
+            ->createQueryBuilder()
+            ->select('u')
+            ->from('App:User', 'u')
+            ->getQuery()
```
Comment (Raresmldvn) [inline src/Repository/UserRepository.php]: this should be the last statement in the list

File: src/Entity/Building.php (hunk 4 above)
Comment (marceldobrica) [reply src/Entity/Building.php]: Solved.

File: src/Controller/Admin/BuildingsController.php (hunk 1 above)
Comment (marceldobrica) [reply src/Controller/Admin/BuildingsController.php]: I have refactorized all routes and route names in admin area. My console debug:router is:
```
  admin_buildings                                GET        ANY      ANY    /admin/buildings                   
  admin_buildings_add                            GET|POST   ANY      ANY    /admin/buildings/add               
  admin_buildings_delete                         GET|POST   ANY      ANY    /admin/buildings/delete/{id}       
  admin_buildings_update                         GET|POST   ANY      ANY    /admin/buildings/update/{id}       
  admin_login                                    GET|POST   ANY      ANY    /admin/login                       
  admin_logout                                   ANY        ANY      ANY    /admin/logout                      
  admin                                          GET        ANY      ANY    /admin                             
  admin_programme                                GET        ANY      ANY    /admin/programmes                  
  admin_programme_add                            GET|POST   ANY      ANY    /admin/programmes/add              
  admin_programme_delete                         GET|POST   ANY      ANY    /admin/programmes/delete/{id}      
  admin_programme_update                         GET|POST   ANY      ANY    /admin/programmes/update/{id}      
  admin_rooms                                    GET        ANY      ANY    /admin/rooms                       
  admin_rooms_add                                GET|POST   ANY      ANY    /admin/rooms/add                   
  admin_rooms_delete                             GET|POST   ANY      ANY    /admin/rooms/delete/{id}           
  admin_rooms_update                             GET|POST   ANY      ANY    /admin/rooms/update/{id}           
  admin_users                                    GET        ANY      ANY    /admin/users                       
  admin_users_add                                GET|POST   ANY      ANY    /admin/users/add                   
  admin_users_delete                             GET|POST   ANY      ANY    /admin/users/delete/{id}           
  admin_users_update                             GET|POST   ANY      ANY    /admin/users/update/{id}   
```

File: src/Controller/Admin/BuildingsController.php (hunk 2 above)
Comment (marceldobrica) [reply src/Controller/Admin/BuildingsController.php]: Changed to info

File: src/Form/Type/BuildingType.php (hunk 5 above)
Comment (marceldobrica) [reply src/Form/Type/BuildingType.php]: Types are not mandatory as I have configured options to use entity types.... but sometimes are useful especially for formatting. I'll add. Most of the time default values are ok.

File: src/Form/Type/ProgrammeType.php (hunk 6 above)
Comment (marceldobrica) [reply src/Form/Type/ProgrammeType.php]: Types are not mandatory as I have configured options to use entity types.... but sometimes are useful especially for formatting. I'll add. Most of the time default values are ok or I have not figured out what values are better.

**Pull request merged**